### PR TITLE
WINC-516: Add git submodules for Cachito support

### PR DESCRIPTION
**Pull request by mansikulkarni96**
This PR adds git submodules for WMCB, Kubernetes, ovn-kubernetes, CNI plugins
to reference them as subdirectories within wmco repository. This is a requirement to add
Cachito support with git submodules.

### CODE REVIEW

File: docs/HACKING.md [hunk 1]
```diff
@@ -139,4 +139,16 @@ opm index add --from-index $INDEX_REPOSITORY:$INDEX_TAG
...
+## Updating Git submodules
+This project contains git submodules for four components:
+1. windows-machine-config-bootstrapper
+2. kubernetes
+3. ovn-kubernetes
+4. containernetworking-plugins
+
+To update the above git submodules use:
+```shell script
+git submodule update --recursive
```
Comment (aravindhp) [inline docs/HACKING.md:153]: Please add commands for individual sub module updates too.

**Review by aravindhp — CHANGES_REQUESTED** [review]
Nice work, @mansikulkarni96 

Please update the Dockerfiles to use the sub-modules and not pull from other repos.

File: build/Dockerfile [hunk 2]
```diff
@@ -16,29 +16,25 @@ ENV PATH=${PATH}:/usr/local/go/bin
...
-WORKDIR windows-machine-config-bootstrapper
+COPY windows-machine-config-bootstrapper .
 RUN make build
 
 # Build hybrid-overlay
 WORKDIR /build/
-RUN git clone --single-branch --branch release-4.6 https://github.com/openshift/ovn-kubernetes/
-WORKDIR ovn-kubernetes/go-controller/
+COPY ovn-kubernetes .
+WORKDIR go-controller/
```
Comment (sebsoto) [inline build/Dockerfile:25]: Shouldnt this be WORKDIR `ovn-kubernetes/go-controller` ?

File: build/Dockerfile.ci [hunk 3]
```diff
@@ -21,39 +21,29 @@ RUN echo "4a7fa60f323ee1416a4b1425aefc37ea359e9d64df19c326a58953a97ad41ea5  go.t
 RUN tar -C /usr/local -xzf go.tar.gz
 ENV PATH=${PATH}:/usr/local/go/bin
 
+# Build WMCO
+# The source here corresponds to the code in the PR and is placed here by the CI infrastructure.
+RUN mkdir /build/windows-machine-config-operator
+WORKDIR /build/windows-machine-config-operator
+COPY . .
+RUN make build
+
```
Comment (sebsoto) [inline build/Dockerfile.ci:30]: Why is this being moved?

**Review by sebsoto — CHANGES_REQUESTED** [review]

File: build/Dockerfile.ci (hunk 3 above)
Comment (mansikulkarni96) [reply build/Dockerfile.ci:30]: moving the copying of WMCO before building submodule binaries since `COPY . .` will copy all the submodules  and we would only need to change the directory and build them.

File: build/Dockerfile (hunk 2 above)
Comment (aravindhp) [reply build/Dockerfile:25]: How did `operator-sdk build` work with `go-controller`?

File: build/Dockerfile [hunk 4]
```diff
@@ -16,29 +16,25 @@ ENV PATH=${PATH}:/usr/local/go/bin
...
 RUN make windows
 
 # Build Kubernetes node binaries
 WORKDIR /build/
-RUN git clone --branch release-4.6 https://github.com/openshift/kubernetes
-WORKDIR /build/kubernetes
-# TODO: Checking out commit before go 1.15 became a requirement, revert checkout in https://issues.redhat.com/browse/WINC-460
-RUN git checkout f5121a6a6a02ddfafd2bfbf5201b092dc5097ab0
```
Comment (aravindhp) [inline build/Dockerfile:34]: How are you pinning to this commit with the sub-module?

File: build/Dockerfile [hunk 5]
```diff
@@ -16,29 +16,25 @@ ENV PATH=${PATH}:/usr/local/go/bin
...
-RUN git clone --single-branch --branch release-4.6 https://github.com/openshift/ovn-kubernetes/
-WORKDIR ovn-kubernetes/go-controller/
+COPY ovn-kubernetes .
+WORKDIR go-controller/
 RUN make windows
 
 # Build Kubernetes node binaries
 WORKDIR /build/
-RUN git clone --branch release-4.6 https://github.com/openshift/kubernetes
-WORKDIR /build/kubernetes
```
Comment (aravindhp) [inline build/Dockerfile:32]: How is `operator-sdk build` working without this?

**Review by aravindhp — CHANGES_REQUESTED** [review]
Thanks for working on this, @mansikulkarni96. Please address my comments.

File: build/Dockerfile (hunk 2 above)
Comment (mansikulkarni96) [reply build/Dockerfile:25]: @aravindhp here I wasn't creating directories inside build, hence COPY ovn-kubernetes . copies all contents inside of ovn-kuberenetes to /build, since the WORKDIR prior to this step was /build/ it was working fine with WORKDIR being set to /go-controller. Pushing changes now to create those directories and then copy the contents.

File: build/Dockerfile (hunk 5 above)
Comment (mansikulkarni96) [reply build/Dockerfile:32]: I do not understand that, since the WORKDIR is /build/ the makefile was accessible

File: build/Dockerfile (hunk 4 above)
Comment (mansikulkarni96) [reply build/Dockerfile:34]: The submodule added shows the commit it is pinned to, one we add a submodule we need to checkout to that commit by cd into the submodule.

Comment (aravindhp) [reply build/Dockerfile:34]: Please add the commands you executed to add all the sub-modules in the commit message.

File: build/Dockerfile [hunk 6]
```diff
@@ -15,31 +15,30 @@ RUN tar -C /usr/local -xzf go.tar.gz
 ENV PATH=${PATH}:/usr/local/go/bin
 
 # Build WMCB
-WORKDIR /build/
-RUN git clone --single-branch --branch release-4.6 https://github.com/openshift/windows-machine-config-bootstrapper.git
-WORKDIR windows-machine-config-bootstrapper
+RUN mkdir /build/windows-machine-config-bootstrapper
```
Comment (aravindhp) [inline build/Dockerfile:18]: Why is this needed? [WORKDIR](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#workdir) should create the directory.

**Review by aravindhp — CHANGES_REQUESTED** [review]

File: docs/HACKING.md [hunk 7]
```diff
@@ -139,4 +139,26 @@ opm index add --from-index $INDEX_REPOSITORY:$INDEX_TAG
 and removed from it:
 ```shell script
 opm index rm --from-index $INDEX_REPOSITORY:$INDEX_TAG
+```
+
+## Updating Git submodules
+This project contains git submodules for four components:
+1. windows-machine-config-bootstrapper
+# TODO: Check out to latest commit on release-4.6 in https://issues.redhat.com/browse/WINC-460
```
Comment (PratikMahajan) [inline docs/HACKING.md:147]: `#` in markdown will render it as a heading, need to remove this

File: docs/HACKING.md [hunk 8]
```diff
@@ -139,4 +139,26 @@ opm index add --from-index $INDEX_REPOSITORY:$INDEX_TAG
...
+2. kubernetes has been pinned to a commit before go 1.15 became a requirement using the commands below. 
+```shell script
+$ cd kubernetes
+$ git checkout f5121a6a6a02ddfafd2bfbf5201b092dc5097ab0
+```
+3. ovn-kubernetes
+4. containernetworking-plugins
+
+To update all git submodules use:
```
Comment (PratikMahajan) [inline docs/HACKING.md:156]: You can add a make target as well to update all submodules, considering that we will be updating it frequently.

File: build/Dockerfile (hunk 6 above)
Comment (mansikulkarni96) [reply build/Dockerfile]: yes, added this change

File: build/Dockerfile [hunk 9]
```diff
@@ -15,31 +15,26 @@ RUN tar -C /usr/local -xzf go.tar.gz
 ENV PATH=${PATH}:/usr/local/go/bin
 
 # Build WMCB
-WORKDIR /build/
-RUN git clone --single-branch --branch release-4.6 https://github.com/openshift/windows-machine-config-bootstrapper.git
-WORKDIR windows-machine-config-bootstrapper
+WORKDIR /build/windows-machine-config-bootstrapper
```
Comment (aravindhp) [inline build/Dockerfile:18]: Nit: I suggest using the same WORKDIR across Dockerfiles

**Review by aravindhp — APPROVED** [review]
/approve

File: docs/HACKING.md [hunk 10]
```diff
@@ -139,4 +139,26 @@ opm index add --from-index $INDEX_REPOSITORY:$INDEX_TAG
 and removed from it:
 ```shell script
 opm index rm --from-index $INDEX_REPOSITORY:$INDEX_TAG
+```
+
+## Updating Git submodules
```
Comment (aravindhp) [inline docs/HACKING.md:144]: I should have caught this before. I realize this is for updating the submodules after a pull. What we need to add here is how to actually update the submodules from it's respective upstream.
